daffodil-codegen-c/src/main/resources/org/apache/daffodil/codegen/c/files/libruntime/validators.c

```diff
@@ -38,6 +38,32 @@ validate_array_bounds(const char *name, size_t count, size_t minOccurs, size_t m
     }
 }
 
+// Validate element or sequence's dfdl:assert expression returns true
+
+void
+validate_dfdl_assert(bool assertion, const char *expression, const char *failure_type,
+                     ParserOrUnparserState *pu)
+{
+    // The boolean expression that determines if the assertion is true
+    // is evaluated in generated code and the result passed here, as
+    // well as failure_type which determines whether to issue a diagnostic
+    // which is printed later or an error which terminates processing now.
+    if (!assertion && failure_type[0] == 'r')
```

Comment:
Thoughts on converting `failure_type` to a C enum instead of checking the first letter of a string? This seems a bit fragile.

Or maybe replace the `failure_type` parameter with a `bool recoverable` and then pass in `assert.failureType == FailureType.RecoverableError`?